feat: add go-libp2p in 2022 blog entry (#33)
libp2p/blog: src/_blog/go-libp2p-2022.md

---
tags:
  - libp2p
  - Go
title: go-libp2p in 2022
description:
date: 2023-01-11
permalink: ''
translationKey: ''
header_image: /
author: Prithvi Shahi
---

> 2022 has been a year full of accomplishments for the libp2p project. Today we'd like to catch you up on advancements made within [go-libp2p](https://github.com/libp2p/go-libp2p). From exciting developments like shipping new features, to necessary work like organizing code, 2022 had it all. Overall we had seven releases ranging from [v0.18.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.18.0) to [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0), with a number of patch releases in between.

Reviewer: ```suggestion
2022 has been a year full of accomplishments for the libp2p project. Today we'd like to catch you up on advancements made within [go-libp2p](https://github.com/libp2p/go-libp2p). From exciting developments like shipping new features, to necessary work like organizing code, 2022 had it all. Overall we had seven releases ranging from [v0.18.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.18.0) to [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0), with a number of patch releases in between.
```

> 2022 has been a year full of accomplishments for the libp2p project. Today we'd like to catch you up on advancements made within [go-libp2p](https://github.com/libp2p/go-libp2p). From exciting developments like shipping new features, to necessary work like organizing code, 2022 had it all. Overall we had seven releases ranging from [v0.18.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.18.0) to [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0), with a number of patch releases in between.

Reviewer: ```suggestion
We are excited to share with you all the progress that has been made on [go-libp2p](https://github.com/libp2p/go-libp2p) in 2022. It has been a year full of exciting new features, code organization, and a growing team of talented contributors.
```

<!-- above docs is a blocker for merge -->

### More Efficient Handshakes 🤝

In the second half of 2022, the go-libp2p team began working to decrease time to first byte and increase the speed of connection establishment (for transports that don't have native stream multiplexing i.e., TCP and WebSocket). The first part of this effort was the [Early Muxer Negotiation feature](https://github.com/libp2p/specs/blob/master/connections/inlined-muxer-negotiation.md).

In [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0), go-libp2p added optimized muxer selection via [TLS' ALPN extension](https://github.com/libp2p/specs/blob/master/connections/inlined-muxer-negotiation.md#multiplexer-negotiation-over-tls) and [Noise extensions](https://github.com/libp2p/specs/blob/master/connections/inlined-muxer-negotiation.md#multiplexer-negotiation-over-noise).
This resulted in a net saving one round trip, which may seem minimal but is big waste during connection establishment!

<!-- To learn more about how early muxer negotiation works, please read: https://github.com/libp2p/docs/pull/274 -->
<!-- above docs is a blocker for merge -->

## Project Improvements 🏡

> 

Author: Some of these occurred in the first half of 2022, like adding the Resource Manager

The details of our interoperability testing efforts can be seen in the shared [libp2p/test-plans Roadmap](https://github.com/libp2p/test-plans/blob/master/ROADMAP.md#a-multi-dimensional-testinginterop-visibility).

### Monorepo Consolidation 1️⃣

go-libp2p is a monorepo as of the [v0.22.0 release](https://github.com/libp2p/go-libp2p/releases/tag/v0.22.0).

This effort was initiated to address repo sprawl and consolidate go-libp2p modules; the result was a monorepo and a big quality of life improvement. The go-libp2p maintainers always wanted to address sprawl but it wasn’t possible until started after [lazy module loading was added to Go 1.17](https://go.dev/ref/mod#lazy-loading).
The consolidation started in [v0.18.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.18.0) and finished in v0.22.0 release (where go-libp2p-core was finally migrated.)

### Simplify libp2p Setup Logic with Fx 🦄

> In [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0) go-libp2p included a change to use [Fx](https://github.com/uber-go/fx), a Go dependency injection library.

Reviewer 2: Not really, this is not an outwards-facing change.

It just massively simplified how we construct our transports, and it will enable us to properly start and stop libp2p services in the future (see https://github.com/libp2p/go-libp2p/issues/1993).

The [v0.19.0 release](https://github.com/libp2p/go-libp2p/releases/tag/v0.19.0) enabled AutoRelay to discover nodes running Circuit Relay v2. Support for relay v2 was first added in [late 2021 in v0.16.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.16.0) (which also removed support for relay v1 and added the [Direct Connenction Upgrade through Relay](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) protocol.)

### Contributions to other projects 🧑‍💻

A lot of features and changes in go-libp2p motivated (and necessitated) making contributions to other projects.
Chief among these are [quic-go](https://github.com/lucas-clemente/quic-go), [webtransport-go](https://github.com/quic-go/webtransport-go), and [pion/sctp](https://github.com/pion/sctp) & [pion/datachannel](https://github.com/pion/datachannel).

## Plans for 2023 📅

Our plans for the next year can be found in [our go-libp2p Roadmap](https://github.com/libp2p/go-libp2p/blob/master/ROADMAP.md).
If you have suggestions, please feel free to create a PR or make a suggestion [in this tracking issue](https://github.com/libp2p/go-libp2p/issues/1806).

Key areas go-libp2p of focus are:

> 

Author: Good idea, I'll address this

To learn more about this exciting feature, check out our blog post on [WebTransport in libp2p](https://blog.libp2p.io/2022-12-19-libp2p-webtransport/) and the [WebTransport documentation](https://docs.libp2p.io/concepts/transports/webtransport/).

#### WebRTC (Browser to Server) 🛰️

In addition to WebTransport, the go-libp2p team also began work on enabling the WebRTC transport, in partnership with Little Bear Labs. This new transport will allow for connectivity between go-libp2p server nodes and js-libp2p browser nodes using [js-libp2p-webrtc](https://github.com/libp2p/js-libp2p-webrtc). 

While this feature is still under development and testing, you can check out a proof-of-concept demo of it in action on our [libp2p Day 2022 Recap blog post](https://blog.libp2p.io/2022-11-22-libp2p-day-2022-recap/#why-webrtc). 

We expect to release this feature in Q1 2023.

#### QUIC 🐰

> We also made developments to the existing QUIC implementation in go-libp2p. While go-libp2p has supported QUIC for some time, we upgraded to the latest version, [RFC 9000](https://datatracker.ietf.org/doc/), from the [draft-29 version](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-29) (in [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0)), inspired by the [alpha release of the QUIC implementation in rust-libp2p](https://github.com/libp2p/rust-libp2p/releases/tag/v0.50.0).

Reviewer 2: This is not entirely correct, go-libp2p nodes have supported RFC 9000 for quite a while. It's just that we'd _dial_ draft-29 by default.

The only thing that we really changed was that we're now properly distinguishing between the two versions in their multiaddress.

Author: Updated it a bit.

> We also made developments to the existing QUIC implementation in go-libp2p. In go-libp2p, we maintained support for two versions of QUIC, [RFC 9000](https://datatracker.ietf.org/doc/) and the [draft-29 version](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-29). In [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0), go-libp2p changed two things. The first was to properly distinguish between these two QUIC versions (in their multiaddresses) an/d second was to change the default dial behavior. Previously, go-libp2p nodes would dial draft-29 by default but after v0.24.0, nodes prefer the new QUIC version. This was partly inspired by the [alpha release of the QUIC implementation in rust-libp2p](https://github.com/libp2p/rust-libp2p/releases/tag/v0.50.0).

Reviewer 3: ```suggestion
We also made developments to the existing QUIC implementation in go-libp2p. In go-libp2p, we maintained support for two versions of QUIC, [RFC 9000](https://datatracker.ietf.org/doc/) and the [draft-29 version](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-29). In [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0), go-libp2p changed two things. The first was to properly distinguish between these two QUIC versions (in their multiaddresses) and second was to change the default dial behavior. Previously, go-libp2p nodes would dial draft-29 by default but after v0.24.0, nodes prefer the new QUIC version. This was partly inspired by the [alpha release of the QUIC implementation in rust-libp2p](https://github.com/libp2p/rust-libp2p/releases/tag/v0.50.0).
```

<!-- To learn more about how different versions of QUIC work please read: https://github.com/libp2p/docs/pull/238 -->
<!-- above docs is a blocker for merge -->

### Faster Handshakes 🤝

> In the second half of 2022, the go-libp2p team began working to decrease the TTFB (time to first byte) and increase the speed of connection establishment (for transports that don't have native stream multiplexing like TCP and WebSocket). The first part of this effort was the [Early Muxer Negotiation feature](https://github.com/libp2p/specs/blob/master/connections/inlined-muxer-negotiation.md).

Reviewer 2: Should this link to the docs page instead?

Author: Now that it exists, yes will do

### Interoperability Testing ⚙️

We also began a concerted effort to improve interoperability between go-libp2p and libp2p implementations in other languages, such as Rust, JS, and Nim. With the addition of new transports like WebTransport and WebRTC, it is important to ensure that these different implementations can work together seamlessly.

The details of our interoperability testing efforts can be seen in the shared [libp2p/test-plans Roadmap](https://github.com/libp2p/test-plans/blob/master/ROADMAP.md#a-multi-dimensional-testinginterop-visibility).

### Monorepo Consolidation 1️⃣

**go-libp2p is a monorepo as of the [v0.22.0 release](https://github.com/libp2p/go-libp2p/releases/tag/v0.22.0).**

In attempt to address the go-libp2p repo sprawl and consolidate various modules, a monorepo was created that resulted in a big quality of life improvement for the project. The go-libp2p maintainers always wanted to address the sprawl but it wasn’t possible until [lazy module loading was added to Go 1.17](https://go.dev/ref/mod#lazy-loading).

> The consolidation started in [v0.18.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.18.0) and finished in [v0.22.0 release](https://github.com/libp2p/go-libp2p/releases/tag/v0.22.0) (where go-libp2p-core was finally migrated.)

Reviewer 2: Should we add more motivation here? Something along the lines that this make changes and refactors across the code base a lot easier.

Author: done

### Simplify libp2p Setup Logic with Fx 🦄

In [v0.24.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.24.0) go-libp2p included a change to use [Fx](https://github.com/uber-go/fx), a Go dependency injection library.
This enabled simplyfying the logic necessary to [construct libp2p](https://github.com/libp2p/go-libp2p/pull/1858).

### AutoRelay discovers Circuit Relay v2 🔭

> In [v0.19.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.19.0), we enabled AutoRelay to discover nodes running Circuit Relay v2. Support for relay v2 was first added in [late 2021 in v0.16.0](https://github.com/libp2p/go-libp2p/releases/tag/v0.16.0) (which also removed support for relay v1 and added the [Direct Connection Upgrade through Relay](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) protocol). This improvement allows go-libp2p nodes to discover and connect to other nodes running Circuit Relay v2, improving the overall performance and reliability of the network.

Reviewer 2: I feel we need to explain better why this is needed:
Running circuit v1 was pretty expensive, which is why there were only a few nodes on the network. Users would either manually configure these nodes as static relays, or discover them from the DHT.

With circuit v2 it became cheap to run (limited) relays, and public nodes now start the relay service by default. There's now a massive number of v2 relays on the IPFS network, and they don't advertise their service to the DHT any more. 
There's now _so many_ of these nodes, that by just connecting to a small number of nodes (e.g. by joining the DHT), we're (statistically) guaranteed to connect to some relays, and we can use these relays if we discover that we're behind a NAT.

Author: Great point, will reword and add it in

### Contributions to other projects 🧑‍💻

In addition to the improvements made directly to go-libp2p, we also made a number of contributions to other projects in 2022.

> Chief among them were [quic-go](https://github.com/lucas-clemente/quic-go), [webtransport-go](https://github.com/quic-go/webtransport-go), along with [pion/sctp](https://github.com/pion/sctp) and [pion/datachannel](https://github.com/pion/datachannel). These contributions helped progress network protocol development in go and improve the overall ecosystem.

Reviewer 2: There's an important different between quic-go and webtransport-go on the one side and pion on the other: we are (I am) the author of these libraries.

Author: I glossed over that intentionally haha. I'll add the attribution.

Author: done
